clang/test/CodeGen/cheri/cheri-builtin-offsetof.c

```diff
@@ -0,0 +1,70 @@
+// RV32-NEXT:    [[AND:%.*]] = and i32 [[TMP1]], [[TMP2]]
+// RV32-NEXT:    [[TMP3:%.*]] = call ptr addrspace(200) @llvm.cheri.cap.address.set.i32(ptr addrspace(200) [[TMP0]], i32 [[AND]])
+// RV32-NEXT:    [[TMP4:%.*]] = call i32 @llvm.cheri.cap.address.get.i32(ptr addrspace(200) [[TMP3]])
+// RV32-NEXT:    [[TMP5:%.*]] = mul i32 [[TMP4]], 8
+// RV32-NEXT:    [[TMP6:%.*]] = add i32 8, [[TMP5]]
+// RV32-NEXT:    ret i32 [[TMP6]]
+//
+// RV64-LABEL: define dso_local i64 @off_with_intcap
+// RV64-SAME: (ptr addrspace(200) noundef [[BIT:%.*]]) addrspace(200) #[[ATTR0:[0-9]+]] {
+// RV64-NEXT:  entry:
+// RV64-NEXT:    [[BIT_ADDR:%.*]] = alloca ptr addrspace(200), align 16, addrspace(200)
+// RV64-NEXT:    store ptr addrspace(200) [[BIT]], ptr addrspace(200) [[BIT_ADDR]], align 16
+// RV64-NEXT:    [[TMP0:%.*]] = load ptr addrspace(200), ptr addrspace(200) [[BIT_ADDR]], align 16
+// RV64-NEXT:    [[TMP1:%.*]] = call i64 @llvm.cheri.cap.address.get.i64(ptr addrspace(200) [[TMP0]])
+// RV64-NEXT:    [[TMP2:%.*]] = call i64 @llvm.cheri.cap.address.get.i64(ptr addrspace(200) getelementptr (i8, ptr addrspace(200) null, i64 1))
+// RV64-NEXT:    [[AND:%.*]] = and i64 [[TMP1]], [[TMP2]]
+// RV64-NEXT:    [[TMP3:%.*]] = call ptr addrspace(200) @llvm.cheri.cap.address.set.i64(ptr addrspace(200) [[TMP0]], i64 [[AND]])
+// RV64-NEXT:    [[TMP4:%.*]] = call i64 @llvm.cheri.cap.address.get.i64(ptr addrspace(200) [[TMP3]])
+// RV64-NEXT:    [[TMP5:%.*]] = mul i64 [[TMP4]], 16
+// RV64-NEXT:    [[TMP6:%.*]] = add i64 16, [[TMP5]]
+// RV64-NEXT:    ret i64 [[TMP6]]
+//
+unsigned long off_with_intcap(__uintcap_t bit) {
+  return __builtin_offsetof(struct foo, children[bit & 1]);
+}
+
+
+// RV32-LABEL: define dso_local i32 @off_with_unsigned
+// RV32-SAME: (i32 noundef [[BIT:%.*]]) addrspace(200) #[[ATTR0]] {
+// RV32-NEXT:  entry:
+// RV32-NEXT:    [[BIT_ADDR:%.*]] = alloca i32, align 4, addrspace(200)
+// RV32-NEXT:    store i32 [[BIT]], ptr addrspace(200) [[BIT_ADDR]], align 4
+// RV32-NEXT:    [[TMP0:%.*]] = load i32, ptr addrspace(200) [[BIT_ADDR]], align 4
+// RV32-NEXT:    [[AND:%.*]] = and i32 [[TMP0]], 1
+// RV32-NEXT:    [[TMP1:%.*]] = mul i32 [[AND]], 8
+// RV32-NEXT:    [[TMP2:%.*]] = add i32 8, [[TMP1]]
+// RV32-NEXT:    ret i32 [[TMP2]]
+//
+// RV64-LABEL: define dso_local i64 @off_with_unsigned
+// RV64-SAME: (i32 noundef signext [[BIT:%.*]]) addrspace(200) #[[ATTR0]] {
+// RV64-NEXT:  entry:
+// RV64-NEXT:    [[BIT_ADDR:%.*]] = alloca i32, align 4, addrspace(200)
+// RV64-NEXT:    store i32 [[BIT]], ptr addrspace(200) [[BIT_ADDR]], align 4
+// RV64-NEXT:    [[TMP0:%.*]] = load i32, ptr addrspace(200) [[BIT_ADDR]], align 4
+// RV64-NEXT:    [[AND:%.*]] = and i32 [[TMP0]], 1
+// RV64-NEXT:    [[CONV:%.*]] = zext i32 [[AND]] to i64
+// RV64-NEXT:    [[TMP1:%.*]] = mul i64 [[CONV]], 16
+// RV64-NEXT:    [[TMP2:%.*]] = add i64 16, [[TMP1]]
+// RV64-NEXT:    ret i64 [[TMP2]]
+//
+unsigned long off_with_unsigned(unsigned int bit) {
```

Comment:
off_with_intcap is using unsigned __intcap, but off_with_unsigned is using unsigned int. Would be more consistent to call this off_with_int.